File: Hard/first_missing_positive/first_missing_positive.py
from typing import List

def firstMissingPositive(self, nums: List[int]) -> int:
    n = len(nums)
    for i in range(n):
        if nums[i] <= 0:
            nums[i] = n + 1
    for i in range(n):
        num = abs(nums[i])
        if 1 <= num <= n:
            nums[num - 1] = -abs(nums[num - 1])
    for i in range(n):
        if nums[i] > 0:
            return i + 1
    return n + 1

if __name__ == "__main__":
    nums = [1, 2, 0]
    print("For the input: " + str(nums) + ", the first missing positive is: ", firstMissingPositive(None, nums))  # Output: 3

    nums = [3, 4, -1, 1]
    print("For the input: " + str(nums) + ", the first missing positive is: ", firstMissingPositive(None, nums))  # Output: 2

    nums = [7, 8, 9, 11, 12]
    print("For the input: " + str(nums) + ", the first missing positive is: ", firstMissingPositive(None, nums))  # Output: 1

File: Hard/first_missing_positive/first_missing_positive.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        for(int i = 0; i < n; i++){
            while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]){
                swap(nums[i], nums[nums[i] - 1]);
            }
        }
        for(int i = 0; i < n; i++){
            if(nums[i] != i+1){
                return i+1;
            }
        }
        return n + 1;
    }
};

void printVector(const vector<int>& nums) {
    cout << "[";
    for (size_t i = 0; i < nums.size(); i++) {
        cout << nums[i];
        if (i != nums.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    Solution solution;
    vector<int> nums1 = {1, 2, 0};
    vector<int> nums2 = {3, 4, -1, 1};
    vector<int> nums3 = {7, 8, 9, 11, 12};

    vector<vector<int>> allNums = {nums1, nums2, nums3};
    
    for (auto nums : allNums) {
        printVector(nums);
        cout << " -> first missing positive: " << solution.firstMissingPositive(nums) << endl;
    }
    return 0;
}